igvc_navigation/src/back_up_recovery/back_up_recovery.cpp

```diff
@@ -0,0 +1,100 @@
+{
+}
+
+void BackUpRecovery::initialize(std::string name, tf2_ros::Buffer *, costmap_2d::Costmap2DROS *,
+                                costmap_2d::Costmap2DROS *local_costmap)
+{
+  if (!initialized_)
+  {
+    local_costmap_ = local_costmap;
+
+    // get some parameters from the parameter server
+    ros::NodeHandle private_nh("~");
+
+    private_nh.getParam("/move_base_flex/back_up_recovery/velocity", velocity_);
+    private_nh.getParam("/move_base_flex/back_up_recovery/frequency", frequency_);
+    private_nh.getParam("/move_base_flex/back_up_recovery/threshold_distance",
+                        threshold_distance_);  // distance robot backs up before exiting back up behavior
+    private_nh.getParam("/move_base_flex/back_up_recovery/obstacle_distance",
+                        obstacle_distance_);  // closest the robot will get to an obstacle from behind
+
+    world_model_ = new base_local_planner::CostmapModel(*local_costmap_->getCostmap());
+
+    initialized_ = true;
+  }
+  else
+  {
+    ROS_ERROR("You should not call initialize twice on this object, doing nothing");
+  }
+}
+
+BackUpRecovery::~BackUpRecovery()
+{
+  delete world_model_;
+}
+
+void BackUpRecovery::runBehavior()
+{
+  if (!initialized_)
+  {
+    ROS_ERROR("This object must be initialized before runBehavior is called");
+    return;
+  }
+
+  if (local_costmap_ == nullptr)
+  {
+    ROS_ERROR("The costmap passed to the BackupRecovery object cannot be NULL. Doing nothing.");
+    return;
+  }
+  ROS_WARN("Backup recovery behavior started.");
+
+  ros::Rate r(frequency_);
+  ros::NodeHandle n;
+  ros::Publisher vel_pub = n.advertise<geometry_msgs::Twist>("cmd_vel", 5);
+
+  geometry_msgs::PoseStamped global_pose;
+  local_costmap_->getRobotPose(global_pose);
+
+  double current_yaw = tf2::getYaw(global_pose.pose.orientation);
+  geometry_msgs::Point current_position = global_pose.pose.position;
+  geometry_msgs::Point start_position = current_position;
+  double x = current_position.x - std::cos(current_yaw) * obstacle_distance_;
+  double y = current_position.y - std::sin(current_yaw) * obstacle_distance_;
+  double footprint_cost = world_model_->footprintCost(x, y, current_yaw, local_costmap_->getRobotFootprint(), 0.0, 0.0);
+  bool next_step_valid = footprint_cost >= 0.0;
+
+  geometry_msgs::Twist vel_msg;
+  vel_msg.linear.x = -velocity_;
+  vel_pub.publish(vel_msg);
+
+  while (n.ok() && igvc::get_distance(current_position, start_position) < threshold_distance_ && next_step_valid)
+  {
+    local_costmap_->getRobotPose(global_pose);
+    current_position = global_pose.pose.position;
+    x = current_position.x - std::cos(current_yaw) * obstacle_distance_;
+    y = current_position.y - std::sin(current_yaw) * obstacle_distance_;
+    footprint_cost = world_model_->footprintCost(x, y, current_yaw, local_costmap_->getRobotFootprint(), 0.0, 0.0);
+    next_step_valid = footprint_cost >= 0.0;
```

Comment:
Instead of repeating this logic twice, I think you should just extract this out into a separate method that you call, maybe something like `bool isNextStepValid()` or something.